wen-restart/src/wen_restart.rs

```diff
@@ -1,58 +1,232 @@
+        }
+        write_wen_restart_records(wen_restart_path, progress)?;
+        let elapsed = timestamp().saturating_sub(start);
+        let time_left = GOSSIP_SLEEP_MILLIS.saturating_sub(elapsed);
+        if time_left > 0 {
+            sleep(Duration::from_millis(time_left));
+        }
+    }
     Ok(())
 }
 
+pub fn wait_for_wen_restart(
+    wen_restart_path: &PathBuf,
+    last_vote: VoteTransaction,
+    blockstore: Arc<Blockstore>,
+    cluster_info: Arc<ClusterInfo>,
+    bank_forks: Arc<RwLock<BankForks>>,
+    slots_to_repair_for_wen_restart: Option<Arc<RwLock<Vec<Slot>>>>,
+    wait_for_supermajority_threshold_percent: u64,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let mut progress = read_wen_restart_records(wen_restart_path)?;
+    if progress.state() == RestartState::Init {
+        send_restart_last_voted_fork_slots(
+            last_vote.clone(),
+            blockstore.clone(),
+            cluster_info.clone(),
+            &mut progress,
+        )?;
+        increment_and_write_wen_restart_records(wen_restart_path, &mut progress)?
+    }
+    if progress.state() == RestartState::LastVotedForkSlots {
+        aggregate_restart_last_voted_fork_slots(
+            wen_restart_path,
+            wait_for_supermajority_threshold_percent,
+            cluster_info.clone(),
+            bank_forks.clone(),
+            slots_to_repair_for_wen_restart.clone().unwrap(),
+            &mut progress,
+        )?;
+        increment_and_write_wen_restart_records(wen_restart_path, &mut progress)?
+    }
+    // Place holder to make the code compile and run for now.
+    if progress.state() == RestartState::HeaviestFork {
+        increment_and_write_wen_restart_records(wen_restart_path, &mut progress)?
+    }
+    if progress.state() == RestartState::Done {
+        Ok(())
+    } else {
+        panic!("Moving to unexpected state {:?}", progress.state())
+    }
+}
+
+fn increment_and_write_wen_restart_records(
+    records_path: &PathBuf,
+    new_progress: &mut WenRestartProgress,
+) -> Result<(), Error> {
+    let new_state = match new_progress.state() {
+        RestartState::Init => RestartState::LastVotedForkSlots,
+        RestartState::LastVotedForkSlots => RestartState::HeaviestFork,
+        RestartState::HeaviestFork => RestartState::Done,
+        _ => panic!("Unexpected state {:?}", new_progress.state()),
+    };
+    new_progress.set_state(new_state);
+    write_wen_restart_records(records_path, new_progress)
+}
+
+fn read_wen_restart_records(records_path: &PathBuf) -> Result<WenRestartProgress, Error> {
+    match read(records_path) {
+        Ok(buffer) => {
+            let progress = WenRestartProgress::decode(&mut Cursor::new(buffer))?;
+            info!("read record {:?}", progress);
+            Ok(progress)
+        }
+        Err(e) => {
+            warn!("cannot read record {:?}: {:?}", records_path, e);
+            Ok(WenRestartProgress {
```

Comment:
If we can't read records we proceed running the state machine?

Reply:
This will happen when wen_restart is entered the first time, there is no record to begin with, we need to start the state machine from the beginning.

Reply:
Changed.